usr/share/rear/layout/save/GNU/Linux/210_raid_layout.sh

```diff
@@ -1,109 +1,223 @@
+            # Have the component devices string as "first_component_device,second_component_device,..."
+            test $component_devices && component_devices+=",$component_device" || component_devices="$component_device"
+        done
+    fi
+    test $component_devices && raid_layout_entry+=" devices=$component_devices" || Error "No component devices for RAID $raiddevice"
+
+    # Optional mdadm option settings:
+    test $name && raid_layout_entry+=" name=$name"
+
+    line=( $( grep "Version :" $mdadm_details ) )
+    metadata=${line[2]}
+    test $metadata && raid_layout_entry+=" metadata=$metadata"
+
+    line=( $( grep "UUID :" $mdadm_details ) )
+    uuid=${line[2]}
+    test $uuid && raid_layout_entry+=" uuid=$uuid"
+
+    line=( $( grep "Layout :" $mdadm_details ) )
+    layout=${line[2]}
+    # fix up layout for RAID10:
+    # > near=2,far=1 -> n2
+    if test "$level" = "raid10" ; then
+        OIFS=$IFS
+        IFS=","
+        for param in "$layout" ; do
+            copies=${layout%=*}
+            number=${layout#*=}
+            test "$number" -gt 1 && layout="${copies:0:1}$number"
+        done
+        IFS=$OIFS
+    fi
+    # mdadm can print '-unknown-' for a RAID layout
+    # which got recently (2019-12-02) added to RAID0 (it existed before for RAID5 and RAID6 and RAID10) see
+    # https://git.kernel.org/pub/scm/utils/mdadm/mdadm.git/commit/Detail.c?id=329dfc28debb58ffe7bd1967cea00fc583139aca
+    # so we treat '-unknown-' same as an empty value to avoid that layout/prepare/GNU/Linux/120_include_raid_code.sh
+    # will create a 'mdadm' command in diskrestore.sh like "mdadm ... --layout=-unknown- ..." which would fail
+    # during "rear recover" with something like "mdadm: layout -unknown- not understood for raid0"
+    # see https://github.com/rear/rear/issues/2616
+    test $layout -a '-unknown-' != "$layout" && raid_layout_entry+=" layout=$layout"
+
+    chunksize=$( grep "Chunk Size" $mdadm_details | tr -d " " | cut -d ":" -f "2" | sed -r 's/^([0-9]+).+/\1/')
+    test $chunksize && raid_layout_entry+=" chunk=$chunksize"
+
+    spare_devices=""
+    test $raid_devices && spare_devices=$(( total_devices - raid_devices ))
+    test $spare_devices -gt 0 && raid_layout_entry+=" spare-devices=$spare_devices"
+
+    test $container_size -gt 0 && raid_layout_entry+=" size=$container_size"
+
+    echo "$raid_layout_entry" >>$DISKLAYOUT_FILE
+
+    extract_partitions "$raiddevice" >>$DISKLAYOUT_FILE
+
+done < <( mdadm --detail --scan --config=partitions )
+
+# mdadm is required in the recovery system if disklayout.conf contains at least one 'raid' entry
+# see the create_raid function in layout/prepare/GNU/Linux/120_include_raid_code.sh
+# what program calls are written to diskrestore.sh
+# cf. https://github.com/rear/rear/issues/1963
+grep -q '^raid ' $DISKLAYOUT_FILE && REQUIRED_PROGS+=( mdadm ) || true
+grep -q '^raid ' $DISKLAYOUT_FILE && PROGS+=( mdmon ) || true
```

Comment:
wouldn't it be nicer to use an usual if ... then block?
```sh
if grep -q '^raid ' $DISKLAYOUT_FILE ; then
    REQUIRED_PROGS+=( mdadm )
    PROGS+=( mdmon )
fi
```
It is of course subjective, but when one does need multiple one-liners with the same condition, it may be better to avoid them, because the saving is not worth it.

Reply:
You are right.
The second line was added via
https://github.com/rear/rear/pull/2702

Reply:
Done via
https://github.com/rear/rear/pull/2714/commits/e980b65154c67b82bb69b1ca874fbbb01e102dc1